dmoj/commands/validate.py

```diff
@@ -0,0 +1,173 @@
+            raise InvalidCommandException(f'unknown problem(s) {unknown_problems}')
+        total_fails = 0
+        for problem_id in problem_ids:
+            fails = self.validate_problem(problem_id)
+            if fails:
+                print_ansi(f'Problem #ansi[{problem_id}](cyan|bold) #ansi[failed](red|bold).')
+                total_fails += 1
+            else:
+                print_ansi(f'Problem #ansi[{problem_id}](cyan|bold) passed with flying colours.')
+            print()
+
+        print()
+        print('Test complete.')
+        if total_fails:
+            print_ansi(f'#ansi[A total of {total_fails} problem(s) have invalid input](red|bold)')
+        else:
+            print_ansi('#ansi[All problems validated.](green|bold)')
+
+        return total_fails
+
+    def validate_problem(self, problem_id: str) -> int:
+        print_ansi(f'Validating problem #ansi[{problem_id}](cyan|bold)...')
+
+        problem_root = get_problem_root(problem_id)
+        config = ProblemConfig(ProblemDataManager(problem_root))
+
+        if not config or 'validator' not in config:
+            print_ansi('\t#ansi[Skipped](magenta|bold) - No validator found')
+            return 0
+
+        validator_config = config['validator']
+        language = validator_config['language']
+        if language not in all_executors:
+            print_ansi('\t\t#ansi[Skipped](magenta|bold) - Language not supported')
+            return 0
+        time_limit = validator_config.get('time', env.generator_time_limit)
+        memory_limit = validator_config.get('memory', env.generator_memory_limit)
+        compiler_time_limit = validator_config.get('compiler_time_limit', env.generator_compiler_time_limit)
+        read_feedback_from = validator_config.get('feedback', 'stderr')
+
+        if read_feedback_from not in ('stdout', 'stderr'):
+            print_ansi('\t\t#ansi[Skipped](magenta|bold) - Feedback option not supported')
+            return 0
+
+        if isinstance(validator_config.source, str):
+            filenames = [validator_config.source]
+        elif isinstance(validator_config.source.unwrap(), list):
+            filenames = list(validator_config.source.unwrap())
+        else:
+            print_ansi('\t#ansi[Skipped](magenta|bold) - No validator found')
+            return 0
+
+        filenames = [os.path.abspath(os.path.join(problem_root, name)) for name in filenames]
+        try:
+            executor = compile_with_auxiliary_files(filenames, [], language, compiler_time_limit)
+        except CompileError as compilation_error:
+            error_msg = compilation_error.message or 'compiler exited abnormally'
+            print_ansi('#ansi[Failed compiling validator!](red|bold)')
+            print(error_msg.rstrip())
+            return 1
+
+        problem = Problem(problem_id, time_limit, memory_limit, {})
+
+        grader_class = type('Grader', (problem.grader_class,), {'_generate_binary': lambda *_: None})  # don't compile
```

Comment:
This deserves a comment about what we're doing here. Why are we using our own compilation routine instead of the grader's?

Reply:
Graders don't easily allow for multi-file compilation (but we need the grader object to resolve the testcases, and specifically their order) also there's no guarantee the grader hasn't overwritten the compilation routine to do something weird, so I just thought it would be easier to forcibly turn it off.